File: app/src/main/java/com/github/vase4kin/teamcityapp/buildlog/view/BuildLogViewImpl.java
/*
 * Copyright 2016 Andrey Tolpeev
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.vase4kin.teamcityapp.buildlog.view;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.view.View;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.widget.Button;

import com.github.vase4kin.teamcityapp.R;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import me.zhanghai.android.materialprogressbar.MaterialProgressBar;
import tr.xip.errorview.ErrorView;

/**
 * Impl of {@link BuildLogView}
 */
public class BuildLogViewImpl implements BuildLogView, OnBuildLogViewListener {

    @BindView(R.id.progress_wheel)
    MaterialProgressBar mProgressWheel;
    @BindView(R.id.error_view)
    ErrorView mErrorView;
    @BindView(R.id.web_view)
    WebView mWebView;
    @BindView(R.id.auth_view)
    View mAuthView;
    @BindView(R.id.warning_build_log_view)
    View mWarningView;
    @BindView(R.id.auth_button)
    Button mAuthButton;
    @BindView(R.id.view_build_log_button)
    Button mOpenBuildLogInBrowserButton;

    private Unbinder mUnbinder;

    private final View mView;
    private final BuildLogWebViewClient client;

    public BuildLogViewImpl(View mView, BuildLogWebViewClient client) {
        this.mView = mView;
        this.client = client;
    }

    /**
     * {@inheritDoc}
     */
    @SuppressLint("SetJavaScriptEnabled")
    @Override
    public void initViews(final OnBuildLogLoadListener listener) {
        mUnbinder = ButterKnife.bind(this, mView);

        mErrorView.setImageTint(Color.LTGRAY);
        mErrorView.setRetryListener(new ErrorView.RetryListener() {
            @Override
            public void onRetry() {
                listener.loadBuildLog();
            }
        });
        mAuthButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                listener.onAuthButtonClick();
            }
        });
        mOpenBuildLogInBrowserButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                listener.onOpenBuildLogInBrowser();
            }
        });

        mWebView.getSettings().setJavaScriptEnabled(true);
        mWebView.getSettings().setLayoutAlgorithm(WebSettings.LayoutAlgorithm.NORMAL);
        client.setListener(this);

        mWebView.setWebViewClient(client);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void loadBuildLog(String buildLogUrl) {
        showProgressWheel();
        mWebView.loadUrl(buildLogUrl);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void showAuthView() {
        mAuthView.setVisibility(View.VISIBLE);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void hideAuthView() {
        mAuthView.setVisibility(View.GONE);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void showSslWarningView() {
        mWarningView.setVisibility(View.VISIBLE);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void unBindViews() {
        mUnbinder.unbind();
        client.setListener(null);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void showWebView() {
        mWebView.setVisibility(View.VISIBLE);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void hideWebView() {
        mWebView.setVisibility(View.GONE);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void showProgressWheel() {
        mProgressWheel.setVisibility(View.VISIBLE);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void hideProgressWheel() {
        mProgressWheel.setVisibility(View.GONE);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void showError() {
        mErrorView.setVisibility(View.VISIBLE);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void hideError() {
        mErrorView.setVisibility(View.GONE);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void evaluateScript(String script) {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.KITKAT) {
            mWebView.evaluateJavascript(script, null);
        } else {
            mWebView.loadUrl("javascript:" + script + "void(0);");
        }
    }

}

File: app/src/main/java/com/github/vase4kin/teamcityapp/buildlog/view/BuildLogViewImpl.kt
/*
 * Copyright 2019 Andrey Tolpeev
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.vase4kin.teamcityapp.buildlog.view

import android.annotation.SuppressLint
import android.graphics.Color
import android.view.View
import android.webkit.WebSettings
import android.webkit.WebView
import android.widget.Button
import butterknife.BindView
import butterknife.ButterKnife
import butterknife.Unbinder
import com.github.vase4kin.teamcityapp.R
import me.zhanghai.android.materialprogressbar.MaterialProgressBar
import tr.xip.errorview.ErrorView

/**
 * Impl of [BuildLogView]
 */
class BuildLogViewImpl(private val view: View, private val client: BuildLogWebViewClient) :
    BuildLogView, OnBuildLogViewListener {

    @BindView(R.id.progress_wheel)
    lateinit var progressWheel: MaterialProgressBar
    @BindView(R.id.error_view)
    lateinit var errorView: ErrorView
    @BindView(R.id.web_view)
    lateinit var webView: WebView
    @BindView(R.id.auth_view)
    lateinit var authView: View
    @BindView(R.id.warning_build_log_view)
    lateinit var warningView: View
    @BindView(R.id.auth_button)
    lateinit var authButton: Button
    @BindView(R.id.view_build_log_button)
    lateinit var openBuildLogInBrowserButton: Button

    private lateinit var unbinder: Unbinder

    /**
     * {@inheritDoc}
     */
    @SuppressLint("SetJavaScriptEnabled")
    override fun initViews(listener: OnBuildLogLoadListener) {
        unbinder = ButterKnife.bind(this, view)

        errorView.setImageTint(Color.LTGRAY)
        errorView.setRetryListener { listener.loadBuildLog() }
        authButton.setOnClickListener { listener.onAuthButtonClick() }
        openBuildLogInBrowserButton.setOnClickListener { listener.onOpenBuildLogInBrowser() }

        webView.settings.javaScriptEnabled = true
        webView.settings.layoutAlgorithm = WebSettings.LayoutAlgorithm.NORMAL
        webView.settings.builtInZoomControls = true
        webView.settings.displayZoomControls = false
        client.setListener(this)

        webView.webViewClient = client
    }

    /**
     * {@inheritDoc}
     */
    override fun loadBuildLog(buildLogUrl: String) {
        showProgressWheel()
        webView.loadUrl(buildLogUrl)
    }

    /**
     * {@inheritDoc}
     */
    override fun showAuthView() {
        authView.visibility = View.VISIBLE
    }

    /**
     * {@inheritDoc}
     */
    override fun hideAuthView() {
        authView.visibility = View.GONE
    }

    /**
     * {@inheritDoc}
     */
    override fun showSslWarningView() {
        warningView.visibility = View.VISIBLE
    }

    /**
     * {@inheritDoc}
     */
    override fun unBindViews() {
        unbinder.unbind()
        client.setListener(null)
    }

    /**
     * {@inheritDoc}
     */
    override fun showWebView() {
        if (::webView.isInitialized) {
            webView.visibility = View.VISIBLE
        }
    }

    /**
     * {@inheritDoc}
     */
    override fun hideWebView() {
        if (::webView.isInitialized) {
            webView.visibility = View.GONE
        }
    }

    /**
     * {@inheritDoc}
     */
    override fun showProgressWheel() {
        if (::progressWheel.isInitialized) {
            progressWheel.visibility = View.VISIBLE
        }
    }

    /**
     * {@inheritDoc}
     */
    override fun hideProgressWheel() {
        if (::progressWheel.isInitialized) {
            progressWheel.visibility = View.GONE
        }
    }

    /**
     * {@inheritDoc}
     */
    override fun showError() {
        if (::errorView.isInitialized) {
            errorView.visibility = View.VISIBLE
        }
    }

    /**
     * {@inheritDoc}
     */
    override fun hideError() {
        if (::errorView.isInitialized) {
            errorView.visibility = View.GONE
        }
    }

    /**
     * {@inheritDoc}
     */
    override fun evaluateScript(script: String) {
        if (::webView.isInitialized) {
            webView.evaluateJavascript(script, null)
        }
    }
}
